spek-runtime/common/src/main/kotlin/org/spekframework/spek2/runtime/execution/Discovery.kt

```diff
@@ -3,6 +3,6 @@ package org.spekframework.spek2.runtime.execution
 import org.spekframework.spek2.runtime.scope.Path
 import org.spekframework.spek2.runtime.scope.ScopeImpl
 
-data class DiscoveryRequest(val path: Path)
+data class DiscoveryRequest(val path: Path, val sourceDirs: List<String>)
```

Comment:
Naming leads to confusion (or maybe it’s just me). Can you please give a few words about differences between `Path` and directory path as `String`?

I always thought that `Path` is a way to find out which tests to run in the current classpath. I’ve read the original issue, basically you are trying to optimize the discovery phase by narrowing down the search, which is a good thing. But isn’t it a better approach just to add more functionality to the `Path` class? I. e. `Path.getFile` or something.

Reply:
`Path` is used for filtering (think of XPath), it used to run a specific scope. `/com.example/FooSpec` path will only execute `com.example.FooSpec`. While `sourceDirs` is where the engine searches for specs.

Reply:
I've renamed it to `specDirs` if that helps.

Reply:
My main concern is that `Path` already has enough metadata for narrowing down the search. It has the package name (which can be used with `Reflections.forPackages`) and the class name. Shouldn’t we use that instead of directories? Seems like you’ve introduced `Path` intentionally to avoid any file system-related associations.

BTW, what is the goal of `Path` as URI? I. e. why is it set to be `/package/class/`? As far as I can see it in the source code it is just a package + a class all the time (am I wrong?)

In any case, I think it is possible to provide extra public fields for the JVM `Path`, i. e. `package` and `class` which can be used by the JVM runtime. Similar can be done for Native and JS or whatever. Am I wrong in this?

Reply:
`Path` and `specDirs` are fundamentally different things. Firstly path is used in filtering scopes you only want to run, say you have the following spec:

```kotlin
package com.example

object MySpec: Spek({
  group("some group") {
      test("some test 1") { ... }
      test("some test 2") { ... }
  }
})
```

If you only want to execute `some test 1` you pass in a path: `/com.example/MySpec/some group/some test/`. If you want to run all specs in `com.example.` you pass in `/com.example/`. For running `MySpec` only, you pass in `/com/example/MySpec`.

`specDirs` on the other hand is the location where the engine looks for this specs. So it actually points to an actual file in the file system,  in gradle, for example, it can contain: `$buildDir/classes/kotlin/test`.

Reply:
Looking that way dosn’t bring much optimization then since _most of the time_ there is a single `test` directory in module. 

@artem-zinnatullin, WDYT?

Reply:
I think you are still missing the point, that's what this PR trying to do. Narrow down the discovery to `specDirs` otherwise the whole classpath will be searched for specs (that includes product and third-party classes).

Reply:
Ah, that makes sense now. I’m not sure it was mentioned anywhere.

Can we make `sourceDirs` a first argument then?